File: Fundamentals_of_Software_Design/Python_language/Hash_table/hash_table.py
class _Node:
    __slots__ = ('key', 'data', 'next')
    def __init__(self, key, data):
        self.key = key
        self.data = data
        self.next = None

class HashTable:
    __slots__ = ('size', 'table', 'hashfunc', 'dtor')
    def __init__(self, size, hashfunc=None, dtor=None):
        self.size = size
        self.table = [None] * size
        self.hashfunc = hashfunc or self._jenkins_hash
        self.dtor = dtor

def _jenkins_hash(key):
    hash_val = 0
    for char in key:
        hash_val += ord(char)
        hash_val += (hash_val << 10)
        hash_val ^= (hash_val >> 6)
    hash_val += (hash_val << 3)
    hash_val ^= (hash_val >> 11)
    hash_val += (hash_val << 15)
    return hash_val

def ht_init(ht, size, hf, dtor):
    ht.size = size
    ht.table = [None] * size
    ht.hashfunc = hf or _jenkins_hash
    ht.dtor = dtor

def ht_destroy(ht):
    for i in range(ht.size):
        current = ht.table[i]
        while current:
            next_node = current.next
            if ht.dtor:
                ht.dtor(current.data)
            current = next_node
    ht.table = None
    ht.size = 0

def ht_set(ht, key, data):
    hash_val = ht.hashfunc(key) % ht.size
    current = ht.table[hash_val]
    
    while current:
        if current.key == key:
            if ht.dtor:
                ht.dtor(current.data)
            current.data = data
            return data
        current = current.next
    
    new_node = _Node(key, data)
    new_node.next = ht.table[hash_val]
    ht.table[hash_val] = new_node
    return data

def ht_get(ht, key):
    hash_val = ht.hashfunc(key) % ht.size
    current = ht.table[hash_val]
    
    while current:
        if current.key == key:
            return current.data
        current = current.next
    return None

def ht_has(ht, key):
    return ht_get(ht, key) is not None

def ht_delete(ht, key):
    hash_val = ht.hashfunc(key) % ht.size
    current = ht.table[hash_val]
    prev = None
    
    while current:
        if current.key == key:
            if prev:
                prev.next = current.next
            else:
                ht.table[hash_val] = current.next
            
            if ht.dtor:
                ht.dtor(current.data)
            return
        
        prev = current
        current = current.next

def ht_traverse(ht, f):
    for i in range(ht.size):
        current = ht.table[i]
        while current:
            f(current.key, current.data)
            current = current.next

def ht_resize(ht, new_size):
    old_table = ht.table
    old_size = ht.size
    
    ht.size = new_size
    ht.table = [None] * new_size
    
    for i in range(old_size):
        current = old_table[i]
        while current:
            next_node = current.next
            hash_val = ht.hashfunc(current.key) % new_size
            current.next = ht.table[hash_val]
            ht.table[hash_val] = current
            current = next_node

File: Fundamentals_of_Software_Design/C_language/Hash_table/hash_table.c
#include <stdlib.h>
#include <string.h>
#include <stdio.h>

#include "hash_table.h"

unsigned jenkins_one_at_a_time_hash(char *key) {
    unsigned hash = 0;
    for (; *key; ++key) {
        hash += *key;
        hash += (hash << 10);
        hash ^= (hash >> 6);
    }
    hash += (hash << 3);
    hash ^= (hash >> 11);
    hash += (hash << 15);
    return hash;
}

void ht_init(HashTable *ht, size_t size, HashFunction hf, Destructor dtor) {
    ht->size = size;
    ht->table = calloc(size, sizeof(List*));
    ht->hashfunc = hf ? hf : jenkins_one_at_a_time_hash;
    ht->dtor = dtor;
}

void ht_destroy(HashTable *ht) {
    for (size_t i = 0; i < ht->size; i++) {
        List *current = ht->table[i];
        while (current) {
            List *next = current->next;
            if (ht->dtor) {
                ht->dtor(current->data);
            }
            free(current->key);
            free(current);
            current = next;
        }
    }
    free(ht->table);
    ht->table = NULL;
    ht->size = 0;
}

Pointer ht_set(HashTable *ht, char *key, Pointer data) {
    unsigned hash = ht->hashfunc(key) % ht->size;
    List *current = ht->table[hash];
    while (current) {
        if (strcmp(current->key, key) == 0) {
            Pointer old_data = current->data;
            if (ht->dtor) {
                ht->dtor(old_data);
            }
            current->data = data;
            return old_data;
        }
        current = current->next;
    }
    List *new_node = malloc(sizeof(List));
    new_node->key = strdup(key);
    new_node->data = data;
    new_node->next = ht->table[hash];
    ht->table[hash] = new_node;
    return NULL;
}

Pointer ht_get(HashTable *ht, char *key) {
    unsigned hash = ht->hashfunc(key) % ht->size;
    List *current = ht->table[hash];
    while (current) {
        if (strcmp(current->key, key) == 0) {
            return current->data;
        }
        current = current->next;
    }
    return NULL;
}

int ht_has(HashTable *ht, char *key) {
    return ht_get(ht, key) != NULL;
}

void ht_delete(HashTable *ht, char *key) {
    unsigned hash = ht->hashfunc(key) % ht->size;
    List *current = ht->table[hash];
    List *prev = NULL;
    while (current) {
        if (strcmp(current->key, key) == 0) {
            if (prev) {
                prev->next = current->next;
            }
            else {
                ht->table[hash] = current->next;
            }
            
            if (ht->dtor) {
                ht->dtor(current->data);
            }
            free(current->key);
            free(current);
            return;
        }
        prev = current;
        current = current->next;
    }
}

void ht_traverse(HashTable *ht, void (*f)(char *key, Pointer data)) {
    for (size_t i = 0; i < ht->size; i++) {
        List *current = ht->table[i];
        while (current) {
            f(current->key, current->data);
            current = current->next;
        }
    }
}

void ht_resize(HashTable *ht, size_t new_size) {
    List **new_table = calloc(new_size, sizeof(List*));
    for (size_t i = 0; i < ht->size; i++) {
        List *current = ht->table[i];
        while (current) {
            List *next = current->next;
            unsigned new_hash = ht->hashfunc(current->key) % new_size;
            current->next = new_table[new_hash];
            new_table[new_hash] = current;
            
            current = next;
        }
    }
    free(ht->table);
    ht->table = new_table;
    ht->size = new_size;
}
